lib/Twig/Cache/Filesystem.php

```diff
@@ -36,9 +36,8 @@ public function __construct($directory, $options = 0)
      */
     public function generateKey($name, $className)
     {
-        $hash = hash('sha256', $className);
-
-        return $this->directory.'/'.$hash[0].'/'.$hash[1].'/'.$hash.'.php';
+        // We assume the class name is suffixed by a hash (see Twig_Environment::getTemplateClass) that we can use to balance the directory tree.
+        return $this->directory.'/'.substr($className, -2).'/'.$className.'.php';
```

Comment:
reverted even though I don't fully agree. It can never work for everyone as it depends on the filesystem and the number of templates you have. So whatever workaround you add it can always be the wrong one depending on who is using it.